internal/cmd/agent.go

```diff
@@ -143,23 +152,41 @@ func writeAgentConfig(pid int) error {
 	return nil
 }
 
+func maybeCancel(msg string, err error, count *counter, cancel context.CancelFunc) {
+	count.count++
+	logging.L.Error().Err(err).Int("count", count.count).Int("threshold", count.threshold).Msg(msg)
+	if count.count > count.threshold {
+		cancel()
+	}
+}
+
 // syncKubeConfig updates the local kubernetes configuration from Infra grants
 func syncKubeConfig(ctx context.Context, cancel context.CancelFunc) {
-	client, err := defaultAPIClient()
-	if err != nil {
-		logging.Errorf("api client: %v\n", err)
+	start := time.Now()
+	count, ok := ctx.Value(countCtxKey).(*counter)
+	if !ok {
+		logging.L.Error().Msg("failed to get counter")
 		cancel()
 	}
 
+	client, err := defaultAPIClient()
+	if err != nil {
+		maybeCancel("failed to get client", err, count, cancel)
+		return
+	}
 	client.Name = "agent"
 
 	user, destinations, grants, err := getUserDestinationGrants(client)
 	if err != nil {
-		logging.Errorf("agent failed to get user destination grants: %v\n", err)
-		cancel()
+		maybeCancel("failed to get grants", err, count, cancel)
+		return
 	}
+
 	if err := writeKubeconfig(user, destinations, grants); err != nil {
-		logging.Errorf("agent failed to update kube config: %v\n", err)
-		cancel()
+		maybeCancel("failed to update kubeconfig", err, count, cancel)
```

Comment:
I wonder if this could leave the kube config in an unexpected state

Reply:
You mean if it was interrupted while modifying the kubeconfig? That shouldn't happen. `writeKubeconfig` writes to a temporary file then does an atomic move.

On closer inspection, `writeKubeconfig` may not need retry since any error is unlikely to be temporary